torch/csrc/autograd/record_function_ops.cpp

```diff
@@ -67,36 +90,71 @@ c10::intrusive_ptr<c10::ivalue::Future> _call_end_callbacks_on_fut(
       };
   // Define a future that completes after the profiling callbacks are run.
   auto profiledFut = fut->then(at::wrapPropagateTLSState(
-      futureProfilingFunc),
+      std::move(futureProfilingFunc)),
       fut->elementType()
       );
   return profiledFut;
 }
 
-// Internal only, do not use directly, use Python's record_function()
-TORCH_LIBRARY_FRAGMENT(profiler, m) {
-    m.def("_record_function_enter(str name, str? args=None) -> Tensor", &record_function_enter);
-    m.def("_record_function_exit", &record_function_exit);
+// Legacy signature using cpp_custom_type_hack
+c10::intrusive_ptr<c10::ivalue::Future> _call_end_callbacks_on_fut_legacy(
+    const at::Tensor &handle,
+    const c10::intrusive_ptr<c10::ivalue::Future>& fut) {
+  return _call_end_callbacks_on_fut(
+      [handle] () -> at::RecordFunction& {
+        TORCH_INTERNAL_ASSERT(
+            handle.defined(),
+            "Undefined RecordFunction handle. This can happen if the handle is "
+            "not correctly persisted and is destroyed before the future is "
+            "realized.");
+
+        return getRecordFunctionFromTensor(handle);
+      },
+      fut
+    );
 }
 
-// Needed to register JIT operator in operator registry below
-c10::AliasAnalysisKind aliasAnalysisFromSchema() {
-  return c10::AliasAnalysisKind::FROM_SCHEMA;
+// New signature using custom_class
+c10::intrusive_ptr<c10::ivalue::Future> _call_end_callbacks_on_fut_new(
+    const c10::intrusive_ptr<PythonRecordFunction> &record,
+    const c10::intrusive_ptr<c10::ivalue::Future>& fut) {
+  return _call_end_callbacks_on_fut(
+      [record] () -> at::RecordFunction& { return record->record; }, fut);
 }
 
-jit::RegisterOperators reg_fut_ops({
-    jit::Operator(
+// Internal only, do not use directly, use Python's record_function()
+TORCH_LIBRARY_FRAGMENT(profiler, m) {
+  m.class_<PythonRecordFunction>("_RecordFunction");
+
+  m.def("_record_function_enter", &record_function_enter_legacy);
```

Comment:
This needs to be `_record_function_enter(str name, str? args=None) -> Tensor`, as there are internal calls which are only providing `name` and failing because the second arg no longer has a default value.